File: sdks/python/salanor_aegis/record.py
"""High-level signed ingest helpers (parity with sdks/typescript/src/record.ts)."""

from __future__ import annotations

from dataclasses import replace
from datetime import datetime, timezone
from typing import Any

from salanor_aegis.enrich_payload import enrich_provenance_payload
from salanor_aegis.ids import new_event_id, new_trace_id
from salanor_aegis.ingest import sign_and_ingest
from salanor_aegis.span import merge_span_payload, new_span_id, span_payload
from salanor_aegis.types import RecordContext, RecordOptions

__all__ = [
    "new_trace_id",
    "new_span_id",
    "span_payload",
    "merge_span_payload",
    "record_trace_start",
    "record_llm_invocation",
    "record_provenance_claim",
    "record_decision",
    "record_data_access",
    "start_span",
    "end_span",
]


def _utc_now_iso() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"


def _apply_enrich(event: dict[str, Any], options: RecordOptions) -> dict[str, Any]:
    if not options.enrich_payload:
        return event
    payload = enrich_provenance_payload(
        payload=dict(event.get("payload") or {}),
        tool_name=event.get("tool_name"),
        action_kind=str(event["action_kind"]),
        policy_id=event.get("policy_id"),
    )
    return {**event, "payload": payload}


def _ingest_event(event: dict[str, Any], options: RecordOptions) -> str:
    span_id = options.span_id
    if span_id:
        payload = dict(event.get("payload") or {})
        event = {
            **event,
            "span_id": span_id,
            "payload": merge_span_payload(span_id, options.span_label, payload),
        }
    enriched = _apply_enrich(event, options)
    result = sign_and_ingest(
        enriched,
        private_key_b64=options.sign.private_key_b64,
        key_id=options.sign.key_id,
        api_base_url=options.ingest.api_base_url,
        ingest_api_key=options.ingest.ingest_api_key,
        idempotency_key=options.ingest.idempotency_key,
    )
    return result.event_id


def record_trace_start(
    ctx: RecordContext,
    *,
    trigger_source: str,
    trigger_detail: str | None = None,
    business_context: str | None = None,
    options: RecordOptions,
) -> str:
    span_id = options.span_id or new_span_id()
    payload = merge_span_payload(
        span_id,
        options.span_label or "Session start",
        {
            "trigger_source": trigger_source,
            "trigger_detail": trigger_detail,
            "business_context": business_context,
            "action": "trace_session_start",
        },
    )
    event = {
        "schema_version": 1,
        "event_id": new_event_id(),
        "organization_id": ctx.organization_id,
        "trace_id": ctx.trace_id,
        "agent_id": ctx.agent_id,
        "key_id": ctx.key_id,
        "emitted_at": _utc_now_iso(),
        "actor_type": "agent",
        "actor_principal": ctx.actor_principal,
        "action_kind": "tool_call",
        "policy_decision": "allow",
        "tool_name": "aegis.trace.start",
        "payload": payload,
    }
    return _ingest_event(event, replace(options, span_id=span_id))


def record_llm_invocation(
    ctx: RecordContext,
    *,
    tool_name: str,
    purpose: str,
    prompt_preview: str,
    response_preview: str,
    data_touched: list[str] | None = None,
    parent_event_id: str | None = None,
    payload: dict[str, Any] | None = None,
    options: RecordOptions,
) -> str:
    base_payload: dict[str, Any] = {
        "purpose": purpose,
        "prompt_preview": prompt_preview,
        "response_preview": response_preview,
        "data_touched": data_touched or [],
        **(payload or {}),
    }
    span_id = options.span_id
    if span_id:
        base_payload = merge_span_payload(span_id, options.span_label, base_payload)

    event: dict[str, Any] = {
        "schema_version": 1,
        "event_id": new_event_id(),
        "organization_id": ctx.organization_id,
        "trace_id": ctx.trace_id,
        "agent_id": ctx.agent_id,
        "key_id": ctx.key_id,
        "emitted_at": _utc_now_iso(),
        "actor_type": "agent",
        "actor_principal": ctx.actor_principal,
        "action_kind": "llm_invocation",
        "policy_decision": "allow",
        "tool_name": tool_name,
        "payload": base_payload,
    }
    if parent_event_id:
        event["parent_event_id"] = parent_event_id
    return _ingest_event(event, options)


def record_provenance_claim(
    ctx: RecordContext,
    *,
    claim: str,
    authority: str,
    subject_event_id: str | None = None,
    business_context: str | None = None,
    options: RecordOptions,
) -> str:
    event = {
        "schema_version": 1,
        "event_id": new_event_id(),
        "organization_id": ctx.organization_id,
        "trace_id": ctx.trace_id,
        "agent_id": ctx.agent_id,
        "key_id": ctx.key_id,
        "emitted_at": _utc_now_iso(),
        "actor_type": "agent",
        "actor_principal": ctx.actor_principal,
        "action_kind": "provenance_claim",
        "policy_decision": "allow_retro_audit",
        "tool_name": "aegis.provenance.claim",
        "payload": {
            "claim": claim,
            "authority": authority,
            "subject_event_id": subject_event_id,
            "business_context": business_context,
            "action": "provenance_assertion",
        },
    }
    return _ingest_event(event, options)


def record_decision(
    ctx: RecordContext,
    *,
    decision: str,
    rationale: str,
    alternatives: list[str] | None = None,
    parent_event_id: str | None = None,
    options: RecordOptions,
) -> str:
    event: dict[str, Any] = {
        "schema_version": 1,
        "event_id": new_event_id(),
        "organization_id": ctx.organization_id,
        "trace_id": ctx.trace_id,
        "agent_id": ctx.agent_id,
        "key_id": ctx.key_id,
        "emitted_at": _utc_now_iso(),
        "actor_type": "agent",
        "actor_principal": ctx.actor_principal,
        "action_kind": "decision",
        "policy_decision": "allow",
        "tool_name": "aegis.agent.decision",
        "payload": {
            "decision": decision,
            "rationale": rationale,
            "alternatives": alternatives or [],
            "action": "agent_decision",
        },
    }
    if parent_event_id:
        event["parent_event_id"] = parent_event_id
    return _ingest_event(event, options)


def record_data_access(
    ctx: RecordContext,
    *,
    operation: str,
    resource: str,
    fields: list[str] | None = None,
    classification: str | None = None,
    parent_event_id: str | None = None,
    options: RecordOptions,
) -> str:
    if operation not in ("read", "write"):
        raise ValueError("operation must be 'read' or 'write'")
    event: dict[str, Any] = {
        "schema_version": 1,
        "event_id": new_event_id(),
        "organization_id": ctx.organization_id,
        "trace_id": ctx.trace_id,
        "agent_id": ctx.agent_id,
        "key_id": ctx.key_id,
        "emitted_at": _utc_now_iso(),
        "actor_type": "agent",
        "actor_principal": ctx.actor_principal,
        "action_kind": "data_access",
        "policy_decision": "allow",
        "tool_name": f"aegis.data.{operation}",
        "payload": {
            "operation": operation,
            "resource": resource,
            "fields": fields or [],
            "classification": classification,
            "action": f"{operation}_{resource}",
        },
    }
    if parent_event_id:
        event["parent_event_id"] = parent_event_id
    return _ingest_event(event, options)


def start_span(
    ctx: RecordContext,
    *,
    label: str,
    parent_span_id: str | None = None,
    options: RecordOptions,
) -> str:
    span_id = options.span_id or new_span_id()
    event: dict[str, Any] = {
        "schema_version": 1,
        "event_id": new_event_id(),
        "organization_id": ctx.organization_id,
        "trace_id": ctx.trace_id,
        "agent_id": ctx.agent_id,
        "key_id": ctx.key_id,
        "emitted_at": _utc_now_iso(),
        "actor_type": "agent",
        "actor_principal": ctx.actor_principal,
        "action_kind": "tool_call",
        "policy_decision": "allow",
        "tool_name": "aegis.span.start",
        "span_id": span_id,
        "payload": {
            "span_id": span_id,
            "span_label": label,
            "parent_span_id": parent_span_id,
            "action": "span_start",
        },
    }
    if parent_span_id:
        event["parent_span_id"] = parent_span_id
    _ingest_event(event, replace(options, span_id=span_id, span_label=label))
    return span_id


def end_span(
    ctx: RecordContext,
    span_id: str,
    options: RecordOptions,
) -> str:
    event = {
        "schema_version": 1,
        "event_id": new_event_id(),
        "organization_id": ctx.organization_id,
        "trace_id": ctx.trace_id,
        "agent_id": ctx.agent_id,
        "key_id": ctx.key_id,
        "emitted_at": _utc_now_iso(),
        "actor_type": "agent",
        "actor_principal": ctx.actor_principal,
        "action_kind": "tool_call",
        "policy_decision": "allow",
        "tool_name": "aegis.span.end",
        "span_id": span_id,
        "payload": {"span_id": span_id, "action": "span_end"},
    }
    return _ingest_event(event, replace(options, span_id=span_id))

File: sdks/typescript/src/record.ts
import { randomUUID } from "node:crypto";
import { enrichProvenancePayload } from "./enrich-payload.js";
import { signAndIngest, type IngestResult } from "./ingest.js";
import { mergeSpanPayload, newSpanId, spanPayload } from "./span.js";
import type { ApsEvent, SignAndIngestOptions, SignOptions } from "./types.js";

export type RecordContext = {
  organizationId: string;
  agentId: string;
  keyId: string;
  traceId: string;
  actorPrincipal: string;
};

export type RecordOptions = {
  sign: SignOptions;
  ingest: SignAndIngestOptions;
  spanId?: string;
  spanLabel?: string;
  enrichPayload?: boolean;
};

function newEventId(): string {
  return `evt_${randomUUID().replace(/-/g, "").slice(0, 24)}`;
}

function applyEnrich(event: ApsEvent, enrich: boolean): ApsEvent {
  if (!enrich) {
    return event;
  }
  return {
    ...event,
    payload: enrichProvenancePayload({
      payload: event.payload,
      toolName: event.tool_name,
      actionKind: event.action_kind,
      policyId: event.policy_id,
    }),
  };
}

async function ingestEvent(
  event: ApsEvent,
  options: RecordOptions,
): Promise<string> {
  const spanId = options.spanId;
  const withSpan: ApsEvent = spanId
    ? {
        ...event,
        span_id: spanId,
        payload: mergeSpanPayload(spanId, options.spanLabel, event.payload),
      }
    : event;
  const result = await signAndIngest(
    applyEnrich(withSpan, options.enrichPayload !== false),
    options.sign,
    options.ingest,
  );
  return result.event_id;
}

/** Opens a trace session (signed). Returns event id. */
export async function recordTraceStart(
  ctx: RecordContext,
  input: {
    triggerSource: string;
    triggerDetail?: string;
    businessContext?: string;
  },
  options: RecordOptions,
): Promise<string> {
  const spanId = options.spanId ?? newSpanId();
  const event: ApsEvent = {
    schema_version: 1,
    event_id: newEventId(),
    organization_id: ctx.organizationId,
    trace_id: ctx.traceId,
    agent_id: ctx.agentId,
    key_id: ctx.keyId,
    emitted_at: new Date().toISOString(),
    actor_type: "agent",
    actor_principal: ctx.actorPrincipal,
    action_kind: "tool_call",
    policy_decision: "allow",
    tool_name: "aegis.trace.start",
    payload: mergeSpanPayload(spanId, options.spanLabel ?? "Session start", {
      trigger_source: input.triggerSource,
      trigger_detail: input.triggerDetail,
      business_context: input.businessContext,
      action: "trace_session_start",
    }),
  };

  return ingestEvent(event, options);
}

/** Records an LLM boundary event (signed + ingested). */
export async function recordLlmInvocation(
  ctx: RecordContext,
  input: {
    toolName: string;
    purpose: string;
    promptPreview: string;
    responsePreview: string;
    dataTouched?: string[];
    parentEventId?: string;
    payload?: Record<string, unknown>;
  },
  options: RecordOptions,
): Promise<string> {
  const spanId = options.spanId;
  const basePayload: Record<string, unknown> = {
    purpose: input.purpose,
    prompt_preview: input.promptPreview,
    response_preview: input.responsePreview,
    data_touched: input.dataTouched ?? [],
    ...input.payload,
  };
  const payload =
    spanId != null
      ? mergeSpanPayload(spanId, options.spanLabel, basePayload)
      : basePayload;

  const event: ApsEvent = {
    schema_version: 1,
    event_id: newEventId(),
    organization_id: ctx.organizationId,
    trace_id: ctx.traceId,
    agent_id: ctx.agentId,
    key_id: ctx.keyId,
    emitted_at: new Date().toISOString(),
    actor_type: "agent",
    actor_principal: ctx.actorPrincipal,
    action_kind: "llm_invocation",
    policy_decision: "allow",
    tool_name: input.toolName,
    parent_event_id: input.parentEventId,
    payload,
  };

  return ingestEvent(event, options);
}

/** Signed provenance claim (standalone APS event). */
export async function recordProvenanceClaim(
  ctx: RecordContext,
  input: {
    claim: string;
    authority: string;
    subjectEventId?: string;
    businessContext?: string;
  },
  options: RecordOptions,
): Promise<string> {
  const event: ApsEvent = {
    schema_version: 1,
    event_id: newEventId(),
    organization_id: ctx.organizationId,
    trace_id: ctx.traceId,
    agent_id: ctx.agentId,
    key_id: ctx.keyId,
    emitted_at: new Date().toISOString(),
    actor_type: "agent",
    actor_principal: ctx.actorPrincipal,
    action_kind: "provenance_claim",
    policy_decision: "allow_retro_audit",
    tool_name: "aegis.provenance.claim",
    payload: {
      claim: input.claim,
      authority: input.authority,
      subject_event_id: input.subjectEventId,
      business_context: input.businessContext,
      action: "provenance_assertion",
    },
  };
  return ingestEvent(event, options);
}

/** Agent decision point (planner / router / branch). */
export async function recordDecision(
  ctx: RecordContext,
  input: {
    decision: string;
    rationale: string;
    alternatives?: string[];
    parentEventId?: string;
  },
  options: RecordOptions,
): Promise<string> {
  const event: ApsEvent = {
    schema_version: 1,
    event_id: newEventId(),
    organization_id: ctx.organizationId,
    trace_id: ctx.traceId,
    agent_id: ctx.agentId,
    key_id: ctx.keyId,
    emitted_at: new Date().toISOString(),
    actor_type: "agent",
    actor_principal: ctx.actorPrincipal,
    action_kind: "decision",
    policy_decision: "allow",
    tool_name: "aegis.agent.decision",
    parent_event_id: input.parentEventId,
    payload: {
      decision: input.decision,
      rationale: input.rationale,
      alternatives: input.alternatives ?? [],
      action: "agent_decision",
    },
  };
  return ingestEvent(event, options);
}

/** Data read/write boundary for audit (PII, stores, files). */
export async function recordDataAccess(
  ctx: RecordContext,
  input: {
    operation: "read" | "write";
    resource: string;
    fields?: string[];
    classification?: string;
    parentEventId?: string;
  },
  options: RecordOptions,
): Promise<string> {
  const event: ApsEvent = {
    schema_version: 1,
    event_id: newEventId(),
    organization_id: ctx.organizationId,
    trace_id: ctx.traceId,
    agent_id: ctx.agentId,
    key_id: ctx.keyId,
    emitted_at: new Date().toISOString(),
    actor_type: "agent",
    actor_principal: ctx.actorPrincipal,
    action_kind: "data_access",
    policy_decision: "allow",
    tool_name: `aegis.data.${input.operation}`,
    parent_event_id: input.parentEventId,
    payload: {
      operation: input.operation,
      resource: input.resource,
      fields: input.fields ?? [],
      classification: input.classification,
      action: `${input.operation}_${input.resource}`,
    },
  };
  return ingestEvent(event, options);
}

/** Opens a formal span (DB row + signed span.start event). */
export async function startSpan(
  ctx: RecordContext,
  input: { label: string; parentSpanId?: string },
  options: Omit<RecordOptions, "spanId"> & { spanId?: string },
): Promise<string> {
  const spanId = options.spanId ?? newSpanId();
  const event: ApsEvent = {
    schema_version: 1,
    event_id: newEventId(),
    organization_id: ctx.organizationId,
    trace_id: ctx.traceId,
    agent_id: ctx.agentId,
    key_id: ctx.keyId,
    emitted_at: new Date().toISOString(),
    actor_type: "agent",
    actor_principal: ctx.actorPrincipal,
    action_kind: "tool_call",
    policy_decision: "allow",
    tool_name: "aegis.span.start",
    span_id: spanId,
    parent_span_id: input.parentSpanId,
    payload: {
      span_id: spanId,
      span_label: input.label,
      parent_span_id: input.parentSpanId,
      action: "span_start",
    },
  };
  await ingestEvent(event, { ...options, spanId, spanLabel: input.label });
  return spanId;
}

/** Closes a span (signed span.end event). */
export async function endSpan(
  ctx: RecordContext,
  spanId: string,
  options: RecordOptions,
): Promise<string> {
  const event: ApsEvent = {
    schema_version: 1,
    event_id: newEventId(),
    organization_id: ctx.organizationId,
    trace_id: ctx.traceId,
    agent_id: ctx.agentId,
    key_id: ctx.keyId,
    emitted_at: new Date().toISOString(),
    actor_type: "agent",
    actor_principal: ctx.actorPrincipal,
    action_kind: "tool_call",
    policy_decision: "allow",
    tool_name: "aegis.span.end",
    span_id: spanId,
    payload: { span_id: spanId, action: "span_end" },
  };
  return ingestEvent(event, { ...options, spanId });
}

export function newTraceId(): string {
  return `trc_${randomUUID().replace(/-/g, "").slice(0, 24)}`;
}

export { newSpanId, spanPayload, type IngestResult };
